### PR TITLE
Allow custom cilium registry images

### PR DESCRIPTION
**Pull request by jorge07**
# Motivation

Prepare for new Docker hub restrictions 

# Description

Allow custom images for cilium

### CODE REVIEW

**Comment by olemarkus** [comment]
Technically the registry proxy feature is more applicable for the use case mentioned. However allowing to specify the image registry brings cilium addon inline with other addons that allows configuring this.

/ok-to-test

**Comment by jorge07** [comment]
@olemarkus the goal it's to be able to run you own images. Like in other projects, we sometimes need to contribute and this require put things in production even when are not stable.

File: docs/networking/cilium.md [hunk 1]
```diff
@@ -138,6 +138,16 @@ Once the secret has been created, encryption can be enabled by setting `enableEn
       enableEncryption: true
 ```
 
+#### Custom images
+
+`docker.io/cilium/cilium` it's the default image but there're cases, after Docker Hub T&C changes, where requirements needs another registry. You can set you own setting `image`, `operatorImage` or, `hubbleRelayImage`:
```
Comment (olemarkus) [inline docs/networking/cilium.md:143]: I would not refer to the Docker TOS here.

File: pkg/apis/kops/networking.go [hunk 2]
```diff
@@ -244,7 +244,12 @@ const CiliumIpamEni = "eni"
 type CiliumNetworkingSpec struct {
 	// Version is the version of the Cilium agent and the Cilium Operator.
 	Version string `json:"version,omitempty"`
-
+	// Image specifies the registry and the image name
+	Image string `json:"image,omitempty"`
```
Comment (olemarkus) [inline pkg/apis/kops/networking.go:248]: What you actually specify here is the image repository, not the image (as image also includes the tag)
I would also perhaps use `AgentRepository` to make what this is for more explicit

Comment (jorge07) [reply pkg/apis/kops/networking.go:248]: Way better. Will change

File: docs/networking/cilium.md (hunk 1 above)
Comment (jorge07) [reply docs/networking/cilium.md:143]: Will remove, no problem

File: upup/models/cloudup/resources/addons/networking.cilium.io/k8s-1.12.yaml.template [hunk 3]
```diff
@@ -443,7 +443,7 @@ spec:
           value: {{ . }}
         {{ end }}
 {{ with .Networking.Cilium }}
-        image: "docker.io/cilium/cilium:{{ .Version  }}"
+        image: "{{- if .Image -}}{{ .Image }}{{- else -}}docker.io/cilium/cilium{{- end -}}:{{ .Version }}"
```
[the remark this answers is not in the document]
Comment (jorge07) [reply upup/models/cloudup/resources/addons/networking.cilium.io/k8s-1.12.yaml.template:446]: Much better. Will Refactor

**Comment by jorge07** [comment]
Thanks for the feedback @olemarkus, all applied

File: docs/networking/cilium.md [hunk 4]
```diff
@@ -138,6 +138,16 @@ Once the secret has been created, encryption can be enabled by setting `enableEn
       enableEncryption: true
 ```
 
+#### Custom images
+
+`docker.io/cilium/cilium` it's the default agent image but there're cases where requirements needs another registry. You can set you own registry for `AgentRepository`, `OperatorRepository` or, `HubbleRelayRepository`:
```
Comment (olemarkus) [inline docs/networking/cilium.md:143]:
```suggestion
If you want to use other image repositories than the default, you can specify them like this:
```

File: pkg/apis/kops/v1alpha2/networking.go [hunk 5]
```diff
@@ -242,7 +242,12 @@ type AmazonVPCNetworkingSpec struct {
 type CiliumNetworkingSpec struct {
 	// Version is the version of the Cilium agent and the Cilium Operator.
 	Version string `json:"version,omitempty"`
-
+	// AgentRepository specifies the registry for cilium agent
```
Comment (olemarkus) [inline pkg/apis/kops/v1alpha2/networking.go:245]:
```suggestion
	// AgentRepository specifies the repository for cilium agent
```

File: pkg/apis/kops/v1alpha2/networking.go [hunk 6]
```diff
@@ -242,7 +242,12 @@ type AmazonVPCNetworkingSpec struct {
...
+	AgentRepository string `json:"agentRepository,omitempty"`
+	// OperatorRepository specifies the registry for cilium operator
```
Comment (olemarkus) [inline pkg/apis/kops/v1alpha2/networking.go:247]:
```suggestion
	// OperatorRepository specifies the repository for cilium operator
```

File: pkg/apis/kops/v1alpha2/networking.go [hunk 7]
```diff
@@ -242,7 +242,12 @@ type AmazonVPCNetworkingSpec struct {
...
+	OperatorRepository string `json:"operatorRepository,omitempty"`
+	// HubbleRelayRepository specifies the registry for cilium relay
```
Comment (olemarkus) [inline pkg/apis/kops/v1alpha2/networking.go:249]:
```suggestion
	// HubbleRelayRepository specifies the repository for cilium relay
```

File: pkg/apis/kops/networking.go [hunk 8]
```diff
@@ -244,7 +244,12 @@ const CiliumIpamEni = "eni"
 type CiliumNetworkingSpec struct {
 	// Version is the version of the Cilium agent and the Cilium Operator.
 	Version string `json:"version,omitempty"`
-
+	// AgentRepository specifies the registry for cilium agent
```
Comment (olemarkus) [inline pkg/apis/kops/networking.go:247]:
```suggestion
	// AgentRepository specifies the repository for cilium agent
```

File: pkg/apis/kops/networking.go [hunk 9]
```diff
@@ -244,7 +244,12 @@ const CiliumIpamEni = "eni"
...
+	AgentRepository string `json:"agentRepository,omitempty"`
+	// OperatorRepository specifies the registry for cilium operator
```
Comment (olemarkus) [inline pkg/apis/kops/networking.go:249]:
```suggestion
	// OperatorRepository specifies the repository for cilium operator
```

File: pkg/apis/kops/networking.go [hunk 10]
```diff
@@ -244,7 +244,12 @@ const CiliumIpamEni = "eni"
...
+	OperatorRepository string `json:"operatorRepository,omitempty"`
+	// HubbleRelayRepository specifies the registry for cilium relay
```
Comment (olemarkus) [inline pkg/apis/kops/networking.go:251]:
```suggestion
	// HubbleRelayRepository specifies the repository for cilium relay
```

**Comment by johngmyers** [comment]
Why the per-component duplication of the `spec.assets.containerRegistry` field? If you want to use a private registry that should be configured across the cluster, not on a component-by-component basis.

**Comment by jorge07** [comment]
That sounds ideal but doesn't cover many things. 
Projects can move between registries, and they do, and have it global means all of them should be living in the same one. 
Another consideration are bug fixing that requires run a custom version for some time, this ones will be living elsewhere and you should be able to do it safely, as with the current implementation you can only do it manually and a rolling update overwrites your changes.

**Comment by olemarkus** [comment]
I agree with @jorge07 on this one. usually you would want a custom agent, but leave the other components as is. Also, if you'd like  to use `cilium/cilium-dev` (which you have to in order to use arm right now) you have to specify repository like this.

**Comment by olemarkus** [comment]
/lgtm

But approvers may want you to squash your commits.

**Comment by jorge07** [comment]
Squashed and renamed the original commit to use the proper terms

File: upup/models/cloudup/resources/addons/networking.cilium.io/k8s-1.12-v1.8.yaml.template [hunk 11]
```diff
@@ -886,7 +886,7 @@ spec:
             topologyKey: "kubernetes.io/hostname"
       containers:
         - name: hubble-relay
-          image: "docker.io/cilium/hubble-relay:{{ .Networking.Cilium.Version }}"
+          image: "{{ .HubbleRelayRepository }}:{{ .Networking.Cilium.Version }}"
```
Comment (hakman) [inline upup/models/cloudup/resources/addons/networking.cilium.io/k8s-1.12-v1.8.yaml.template:889]: Out of curiosity, any reason to use `.Networking.Cilium.Version` and not `.Version` here also?

Comment (jorge07) [reply upup/models/cloudup/resources/addons/networking.cilium.io/k8s-1.12-v1.8.yaml.template]: I did a find and replace and missed that, thanks, done

**Comment by johngmyers** [comment]
If you want a custom image then you will likely need to specify a tag as well, not just a repository. The feature is not fit for that use case.

**Comment by jorge07** [comment]
With the current implementation I can specify the tag/version. True that should be the same for the 3 but I can deal with that. Repositories are hardcoded, so that's the main reason. If you see more useful a version for component I can add it too, sounds very reasonable.

**Comment by johngmyers** [comment]
Feedback from the informal Kops meeting of November 12:

* `spec.assets.containerRegistry` and the assets phase are the correct mechanism for addressing the use case of mirroring to a local registry, for example to avoid Docker Hub rate limits.
* Existing API fields for supporting custom images take the entire image path, including image name and tag. If supporting custom images is a requirement, any new fields should take the entire image path for consistency.
* A future API version might replace such image paths with a structure.
* One suggested alternative would be to add a general feature for replacing arbitrary images with custom images. Such a feature should not be conflated with assets phase, which should only address mirroring.

**Comment by jorge07** [comment]
I'll try to get that suggestion done but I'll probably need some guidelines.
Context [in this pr](https://github.com/cilium/cilium/pull/14299)

**Comment by olemarkus** [comment]
I suppose we can close this particular PR now?

We can certainly help out with the alternative suggestions

**Comment by jorge07** [comment]
Let's address this in another PR as will not be even close to this one original idea.

**Pull request closed without merging**